linkerd/distribute/src/params.rs

```diff
@@ -64,46 +67,46 @@ impl<K> Default for Distribution<K> {
 }
 
 impl<K> Distribution<K> {
-    pub fn first_available(keys: impl IntoIterator<Item = K>) -> Self {
-        let keys: Arc<[K]> = keys.into_iter().collect();
+    pub fn first_available(iter: impl IntoIterator<Item = K>) -> Self {
+        let keys = UnweightedKeys::new(iter.into_iter());
         if keys.is_empty() {
             return Self::Empty;
         }
-        Self::FirstAvailable(keys)
+
+        Self::FirstAvailable(Arc::new(keys))
     }
 
     pub fn random_available<T: IntoIterator<Item = (K, u32)>>(
         iter: T,
     ) -> Result<Self, WeightedError> {
-        let (keys, weights): (Vec<_>, Vec<_>) = iter.into_iter().filter(|(_, w)| *w > 0).unzip();
-        if keys.len() < 2 {
-            return Ok(Self::first_available(keys));
+        let weighted_keys = WeightedKeys::new(
+            iter.into_iter()
+                .map(|(key, weight)| WeightedKey { key, weight }),
+        );
+        if weighted_keys.len() < 2 {
+            return Ok(Self::FirstAvailable(Arc::new(
+                weighted_keys.into_unweighted(),
+            )));
         }
-        // Error if the distribution is invalid.
-        let _index = WeightedIndex::new(weights.iter().copied())?;
-        Ok(Self::RandomAvailable(Arc::new(WeightedKeys {
-            keys,
-            weights,
-        })))
+
+        weighted_keys.validate_weights()?;
+        Ok(Self::RandomAvailable(Arc::new(weighted_keys)))
     }
 
-    pub(crate) fn keys(&self) -> &[K] {
+    pub(crate) fn make_svc_from_keys<T>(
```

Comment:
This was another key change. Instead of exposing a slice of keys to create the backend services, this inverts the control to something similar to a combinator to decouple how they keys are iterated over and mapped into services.

Reply:
Done.